inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp

```diff
@@ -382,119 +382,154 @@ class ScaleFactorPerLayer<InferenceEngine::ConcatLayer*> {
             quantData->_dst_quant.scale = quantParams0->_dst_quant.scale;
             quantData->_src_quant.scale = quantParams0->_dst_quant.scale;
             return true;
         }
-        // support only cases when one of input is network input
-        if (infoIn0.isInput() && infoIn1.isInput()) {
-            THROW_GNA_EXCEPTION << "Two Input layers " << in0->name << "and" << in1->name << " has different scales in concat!!! \n";
-        }
 
-        int concatIdxToUpdate = -1;
+        // check if all inputs have the same quant value
+        auto inputLayerCheck = [](InferenceEngine::CNNLayerPtr& inputLayer) {
+            auto info = LayerInfo(inputLayer);
+            return info.isInput();
+        };
 
-        if (infoIn0.isInput()) {
-            sourceQuantParams = quantParams0;
-        } else if (infoIn1.isInput()) {
-            concatIdxToUpdate = 0;
-            sourceQuantParams = quantParams1;
+        GNAPluginNS::QuantizedLayerParams* sourceQuantParams = nullptr;
+        auto firstInputIt = std::find_if(inputLayers.begin(), inputLayers.end(), inputLayerCheck);
+        if (firstInputIt != inputLayers.end()) {
+            auto quantParamsFirst = InferenceEngine::getInjectedData<QuantizedLayerParams>(*firstInputIt);
+            auto nextInputIt = firstInputIt + 1;
+            while ((nextInputIt = std::find_if(nextInputIt, inputLayers.end(), inputLayerCheck)) != inputLayers.end()) {
+                auto quantParamsSecond = InferenceEngine::getInjectedData<QuantizedLayerParams>(*nextInputIt);
+                if (!fp32eq(quantParamsSecond->_dst_quant.scale, quantParamsFirst->_dst_quant.scale)) {
+                    THROW_GNA_EXCEPTION << "Two Input layers " << (*firstInputIt)->name
+                        << " and " << (*nextInputIt)->name << " have different scales in concat!!! \n";
+                }
+            }
         }
 
-        // possible case when some of the concat inputs are free to select scale ex: const->concat<-affine
-        if (quantParams1->_dst_quant.scale == 1.0) {
-            quantParams1->_weights_quant = quantParams0->_dst_quant;
-            quantParams1->_dst_quant     = quantParams0->_dst_quant;
+        // find a source quant value
+        // - 1st candidate - non-activation layer with non-1 scale factor
+        // - 2nd candidate - 1st layer with non-1 scale factor
+        auto sourceLayerCheck = [&fp32eq](InferenceEngine::CNNLayerPtr& inputLayer) {
+            auto quantParams = InferenceEngine::getInjectedData<QuantizedLayerParams>(inputLayer);
+            LayerInfo info(inputLayer);
+            return !info.isActivation() && !fp32eq(quantParams->_dst_quant.scale, 1.0f);
+        };
 
-            sourceQuantParams = quantParams0;
-        }
+        auto sourceLayerIt = std::find_if(inputLayers.begin(), inputLayers.end(), sourceLayerCheck);
+        if (sourceLayerIt == inputLayers.end()) {
```

Comment:
Chosing always the same input layer for scale factor quantization may lead to requantization loop in some cases (it is also the case with current code before your changes) please add some mechanism to change selected input for requantization if this layer is being requantized repeatedly (you can check PR #2080 where map with counter is used for that case, likely not great solution but working)

Reply:
I have pushed a fix for it.